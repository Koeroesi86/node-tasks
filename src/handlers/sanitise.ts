import moxios from 'moxios';

export const sanitiseMoxios = (
  request: ReturnType<typeof moxios.requests.at>
) => ({
  baseUrl: request.config.baseURL,
  url: request.config.url,
  method: request.config.method,
  headers: { ...request.config.headers },
  data: request.config.data,
});
