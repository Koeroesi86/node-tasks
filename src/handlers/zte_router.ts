import axios, { RawAxiosResponseHeaders } from 'axios';

const encode = (obj: Record<string, string>) =>
  Object.entries(obj)
    .map(([k, v]) => encodeURI(k) + '=' + encodeURI(v))
    .join('&');

const parseSID = (setIDCookie: string) =>
  setIDCookie.replace('SID=', '').replace('; PATH=/; HttpOnly', '');

const getSID = (headers: RawAxiosResponseHeaders) =>
  Array.isArray(headers['set-cookie'])
    ? parseSID(headers['set-cookie'].find((c) => c.startsWith('SID=')) || '')
    : parseSID(headers['set-cookie'] ?? '');

export async function callZTERouter({
  router = '',
  user = '',
  password = '',
  action = '',
}) {
  const baseURL = `http://${router}/`;
  const defaultHeaders = {
    'Content-Type': 'application/x-www-form-urlencoded',
    Cookie: '_TESTCOOKIESUPPORT=1',
    Host: router,
    Referer: baseURL,
  };

  const loginResponse = await axios.request({
    baseURL,
    url: '/',
    method: 'POST',
    headers: defaultHeaders,
    data: encode({
      action: 'login',
      Username: user,
      Password: password,
    }),
  });
  const sid = getSID(loginResponse.headers);

  if (!sid) {
    throw Object.assign(new Error('SID not received.'), {
      statusCode: loginResponse.status,
      headers: loginResponse.headers,
    });
  }

  const response = await axios.request({
    baseURL,
    url: '/',
    method: 'POST',
    headers: {
      ...defaultHeaders,
      Cookie: `${defaultHeaders['Cookie']}; SID=${sid}`,
    },
    data: encode({
      action,
    }),
  });

  return response.data;
}
