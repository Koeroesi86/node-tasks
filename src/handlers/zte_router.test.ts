import moxios from 'moxios';
import { callZTERouter } from './zte_router';
import { sanitiseMoxios } from './sanitise';

describe('callZTERouter', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('calls axios', async () => {
    moxios.stubRequest('/', {
      status: 200,
      headers: { 'set-cookie': ['SID=example-token; PATH=/; HttpOnly'] },
      response: {},
    });

    await callZTERouter({
      router: 'example-router',
      user: 'example-user',
      password: 'example-password',
      action: 'exampel-action',
    });

    expect(moxios.requests.count()).toEqual(2);
    expect(sanitiseMoxios(moxios.requests.at(0))).toMatchSnapshot();
    expect(sanitiseMoxios(moxios.requests.at(1))).toMatchSnapshot();
  });
});
