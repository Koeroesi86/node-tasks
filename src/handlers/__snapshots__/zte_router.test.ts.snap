// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callZTERouter calls axios 1`] = `
{
  "baseUrl": "http://example-router/",
  "data": "action=login&Username=example-user&Password=example-password",
  "headers": {
    "Accept": "application/json, text/plain, */*",
    "Content-Type": "application/x-www-form-urlencoded",
    "Cookie": "_TESTCOOKIESUPPORT=1",
    "Host": "example-router",
    "Referer": "http://example-router/",
  },
  "method": "post",
  "url": "/",
}
`;

exports[`callZTERouter calls axios 2`] = `
{
  "baseUrl": "http://example-router/",
  "data": "action=exampel-action",
  "headers": {
    "Accept": "application/json, text/plain, */*",
    "Content-Type": "application/x-www-form-urlencoded",
    "Cookie": "_TESTCOOKIESUPPORT=1; SID=example-token",
    "Host": "example-router",
    "Referer": "http://example-router/",
  },
  "method": "post",
  "url": "/",
}
`;
